rust-runtime/aws-smithy-wasm/src/wasi.rs

```diff
@@ -0,0 +1,345 @@
+            let fut = fut_result?;
+            let response = fut.map(|body| {
+                if body.is_empty() {
+                    SdkBody::empty()
+                } else {
+                    SdkBody::from(body)
+                }
+            });
+            tracing::trace!("WasiHttpConnector: response received {response:?}");
+
+            let sdk_res = Response::try_from(response)
+                .map_err(|err| ConnectorError::other(err.into(), None))?;
+
+            Ok(sdk_res)
+        })
+    }
+}
+
+/// WASI HTTP client containing the options passed to the outgoing_handler
+struct WasiDefaultClient {
+    options: WasiRequestOptions,
+}
+
+impl WasiDefaultClient {
+    /// Create a new WASI HTTP client.
+    fn new(options: WasiRequestOptions) -> Self {
+        Self { options }
+    }
+
+    /// Make outgoing HTTP request in a WASI environment
+    fn handle(&self, req: http::Request<Bytes>) -> Result<http::Response<Bytes>, ConnectorError> {
+        let req =
+            WasiRequest::try_from(req).map_err(|err| ConnectorError::other(err.into(), None))?;
+
+        let res = outgoing_handler::handle(req.0, self.options.clone().0)
+            .map_err(|err| ConnectorError::other(err.into(), None))?;
+
+        // Right now only synchronous calls can be made through WASI, so we subscribe and
+        // block on the FutureIncomingResponse
+        let subscription = res.subscribe();
+        subscription.block();
+
+        //The FutureIncomingResponse .get() method returns a
+        //Option<Result<Result<IncomingResponse, ErrorCode>, ()>>.
+        //The outer Option ensures readiness which we know is Some because we .block() waiting for it
+        //The outer Result is just a singleton enforcer so we can only get the response once
+        //The inner Result indicates whether the HTTP call was sent/received successfully (not the 200 succes of the call)
+        let incoming_res = res
+            .get()
+            .expect("Http response not ready")
+            .expect("Http response accessed more than once")
+            .map_err(|err| ConnectorError::other(err.into(), None))?;
+
+        let response = http::Response::try_from(WasiResponse(incoming_res))
+            .map_err(|err| ConnectorError::other(err.into(), None))?;
+
+        Ok(response)
+    }
+}
+
+/// Wrapper for the WASI RequestOptions type to allow us to impl Clone
+#[derive(Debug)]
+struct WasiRequestOptions(Option<outgoing_handler::RequestOptions>);
+
+impl From<&HttpConnectorSettings> for WasiRequestOptions {
+    fn from(value: &HttpConnectorSettings) -> Self {
+        //The WASI Duration is nanoseconds represented as u64
+        let connect_timeout = value
+            .connect_timeout()
+            .map(|dur| u64::try_from(dur.as_nanos()).unwrap_or(u64::MAX));
+        let read_timeout = value
+            .read_timeout()
+            .map(|dur| u64::try_from(dur.as_nanos()).unwrap_or(u64::MAX));
```

Comment:
If they aren't called the default appears to be `600,000` milliseconds ([set here](https://github.com/bytecodealliance/wasmtime/blob/36fb62ca3bea7dd1456f12ead03084e2ceb48cda/crates/wasi-http/src/http_impl.rs#L23-L33)). But this is set in `wasmtime` and not by the WASI bindings, so it seems to be host dependent. To be consistent with that I can change the `u64::MAX` to match their 600 seconds default.

Reply:
Breaking this down a little more: 

The behavior of `set_*_timeout`:
* `set_*_timeout` is never called - then the default timeout (for all of them) is set to `600` seconds ([code](https://github.com/bytecodealliance/wasmtime/blob/36fb62ca3bea7dd1456f12ead03084e2ceb48cda/crates/wasi-http/src/http_impl.rs#L23-L33)). Note that we don't encounter this in our code since the timeouts are always set based on the `HttpConnectorSettings` values. 
* `set_*_timeout` is called with `None` - the timeout is set to 0 ([code](https://github.com/bytecodealliance/wasi/blob/8c5c76f94a35eb3f4022da2a81612d7aa86a573c/src/bindings.rs#L6069C25-L6072C27))
* `set_*_timeout` is called with `Some(u64)` - timeout is set to the `u64` value

The behavior of `WasiRequestOptions::from(HttpConnectorSettings)`:
* The `HttpConnectorSettings` contains a timeout value (`u128`) less than `u64::MAX` - Then the corresponding timeout value on `outgoing_handler::RequestOptions` is set to that value (cast to a `u64`)
* The `HttpConnectorSettings` contains a timeout value (`u128`) greater than `u64::MAX` - Then the value is clamped to `u64::MAX` and set to that

Reply:
Added comments explaining the clamping in https://github.com/smithy-lang/smithy-rs/pull/3409/commits/c883aad007e0f5929794be8f5370854679a5f213